deploy.sh

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -p jq -i bash
+
+# shellcheck shell=bash
+
+set -euo pipefail
+
+
+# For first deployment to bare server
+if [[ ${1:-} == '--prime' ]]; then
+    shift
+
+    BARE_SERVER=1
+else
+    BARE_SERVER=0
+fi
+
+get() {
+    set +u
+    FROM_ENV="$(eval "echo \$$1")"
+    set -u
+    if [[ ! -z "$FROM_ENV" ]]; then
+        echo "$FROM_ENV"
+    else
+        jq -r ".$1"
+    fi
+}
+
+ensure_set() {
+    if [[ "$(eval "echo \$$1")" == null ]]; then
+        echo "$2 is not set for profile $PROFILE of node $NODE"
+        exit 1
+    fi
+}
+
+deploy_profile() {
+    echo "=== Deploying profile $PROFILE of node $NODE ==="
+
+    ONLY_PROFILE="del(.path) | del(.activate)"
+    MERGED="$(jq "(. | del(.nodes) | del(.hostname) | $ONLY_PROFILE) + (.nodes.\"$NODE\" | del(.profiles) | $ONLY_PROFILE) + (.nodes.\"$NODE\".profiles.\"$PROFILE\" | del(.hostname))" <<< "$JSON")"
+
+    HOST="$(get hostname <<< "$MERGED")"
+    SSH_USER="$(get sshUser <<< "$MERGED")"
+    PROFILE_USER="$(get user <<< "$MERGED")"
+    CLOSURE="$(get path <<< "$MERGED")"
+    ACTIVATE="$(get activate <<< "$MERGED")"
+    FAST_CONNECTION="$(get fastConnection <<< "$MERGED")"
+    EXTRA_SSH_OPTS="$(get sshOpts <<< "$MERGED")"
+    BOOTSTRAP="$(get bootstrap <<< "$MERGED")"
+    AUTO_ROLLBACK="$(get autoRollback <<< "$MERGED")"
+
+    ensure_set HOST hostname
+    ensure_set CLOSURE path
+
+    if [[ "$EXTRA_SSH_OPTS" == null ]]; then
+        EXTRA_SSH_OPTS=""
+    fi
+
+    export NIX_SSHOPTS="${EXTRA_SSH_OPTS}"
+
+    SUDO=""
+
+
+    if [[ "$SSH_USER" == null ]]; then
+        if [[ "$PROFILE_USER" == null ]]; then
+            echo "neither user nor sshUser set for profile $PROFILE of node $NODE"
+            exit 1
+        fi
+        SSH_USER="$USER"
+
+    fi
+    if [[ ! "$PROFILE_USER" == null ]] && [[ ! "$PROFILE_USER" == "$SSH_USER" ]]; then
+        SUDO="sudo -u $PROFILE_USER"
+    fi
+
+    if [[ "$PROFILE_USER" == null ]]; then
+        PROFILE_USER="$SSH_USER"
+    fi
+
+
+    if [[ "$PROFILE_USER" == root ]]; then
+        PROFILE_PATH="/nix/var/nix/profiles/$PROFILE"
+    else
+        PROFILE_PATH="/nix/var/nix/profiles/per-user/$PROFILE_USER/$PROFILE"
+    fi
+
+    if [[ "$FLAKE_SUPPORT" == 1 ]]; then
+        nix build --no-link "$REPO#deploy.nodes.$NODE.profiles.$PROFILE.path"
+    else
+        nix-build "$REPO" -A "deploy.nodes.$NODE.profiles.$PROFILE.path" --no-out-link
+    fi
+
+    if [[ -n ${LOCAL_KEY:-} ]]; then
+        nix sign-paths -r -k "$LOCAL_KEY" "$CLOSURE"
+    fi
+
+    if [[ "$ACTIVATE" == null ]]; then
+        ACTIVATE=true
+    fi
+
+    EXTRA_NIX_COPY_OPTS=""
+
+    if [[ ! "$FAST_CONNECTION" == true ]]; then
+        EXTRA_NIX_COPY_OPTS="$EXTRA_NIX_COPY_OPTS --substitute-on-destination"
+    fi
+
+    set -x
+
+    # shellcheck disable=SC2086
+    nix copy $EXTRA_NIX_COPY_OPTS --no-check-sigs --to "ssh://$SSH_USER@$HOST" "$CLOSURE"
```

Comment:
I find your lack of arrays disturbing. Does `shellcheck` not scream at you for this?

Reply:
You're welcome to rewrite this with arrays, please just check that both `examples` still work after you do.

Reply:
Please, don’t rewrite this with arrays.